File: asteroid-collision/asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        ArrayDeque<Integer> stack = new ArrayDeque<>();
        for (int i = 0; i < asteroids.length; i++) {
            if (stack.isEmpty() || asteroids[i] > 0) {
                stack.offer(asteroids[i]);
            }
            else {
                while(stack.peekLast() != null && stack.peekLast() > 0 && stack.peekLast() < Math.abs(asteroids[i])) {
                    stack.pollLast();
                }

                if (stack.peekLast() == null || stack.peekLast() < 0) {
                    stack.offer(asteroids[i]);
                }
                else if (stack.peekLast() == Math.abs(asteroids[i])) {
                    stack.pollLast();
                }
            }
        }
        int[] answer = new int[stack.size()];
        int i = 0;
        while (!stack.isEmpty()) {
            answer[i++] = stack.poll();
        }
        return answer;
    }
}

File: 735-asteroid-collision/asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = deque()
        for ast in asteroids:
            if ast < 0:
                while stack and stack[-1] > 0 and stack[-1] < -ast:
                    stack.pop()
                if not stack or stack[-1] < -ast:
                    stack.append(ast)
                elif stack[-1] == -ast:
                    stack.pop()
            else:
                stack.append(ast)
        return list(stack)